File: julia/camera.jl
module CameraModule

export Camera, get_ray

include("utils.jl")
include("rays.jl")
using .Utils
using .Rays
using LinearAlgebra

struct Camera
    origin::Point3
    lower_left_corner::Point3
    horizontal::Vec3
    vertical::Vec3
end

function Camera()
    aspect_ratio = 16.0 / 9.0
    viewport_height = 2.0
    viewport_width = aspect_ratio * viewport_height
    focal_length = 1.0

    origin = point3(0, 0, 0)
    horizontal = vec3(viewport_width, 0, 0)
    vertical = vec3(0, viewport_height, 0)
    lower_left_corner = origin - horizontal/2 - vertical/2 - vec3(0, 0, focal_length)

    return Camera(
        origin,
        lower_left_corner,
        horizontal,
        vertical
    )
end

function Camera(lookFrom::Point3, lookAt::Point3, vup::Vec3, vfov::Float64, aspect_ratio::Float64)

    θ = deg2rad(vfov)
    h = Base.tan(θ/2)
    viewport_height = 2.0 * h
    viewport_width = aspect_ratio * viewport_height

    w = normalize(lookFrom - lookAt)
    u = normalize(cross(vup, w))
    v = cross(w, u)
    
    focal_length = 1.0

    origin = copy(lookFrom)
    horizontal = viewport_width * u
    vertical = viewport_height * v
    lower_left_corner = origin - horizontal/2 - vertical/2 - w
     
    return Camera(
        origin,
        lower_left_corner,
        horizontal,
        vertical
    )
end

function get_ray(c::Camera, s::Float64, t::Float64)
    return Ray(c.origin, c.lower_left_corner + s*c.horizontal +  t*c.vertical - c.origin)
end

end

File: rust/src/camera.rs
use crate::ray::Ray;
use crate::vec3::{Point3, Vec3};

pub struct Camera {
    origin: Point3,
    ll_corner: Point3,
    horizontal: Vec3,
    vertical: Vec3,
}

impl Camera {
    pub fn new(lookFrom: Point3, lookAt: Point3, vup: Vec3, vfov: f32, aspect_ratio: f32) -> Self {
        let theta = vfov.to_radians();
        let h = (theta / 2.0).tan();
        let viewport_height = 2.0 * h;
        let viewport_width = aspect_ratio * viewport_height;
        let focal_length = 1.0;

        let w = (lookFrom - lookAt).unit_vector();
        let u = Vec3::cross(&vup, &w).unit_vector();
        let v = Vec3::cross(&w, &u);

        let origin = lookFrom;
        let horizontal = viewport_width * u;
        let vertical = viewport_height * v;
        let ll_corner = origin - horizontal / 2. - vertical / 2. - w;

        Camera {
            origin,
            ll_corner,
            horizontal,
            vertical,
        }
    }

    pub fn get_ray(&self, u: f32, v: f32) -> Ray {
        Ray {
            origin: self.origin,
            direction: self.ll_corner + u * self.horizontal + v * self.vertical - self.origin,
        }
    }
}
